src/hashtable.c

```diff
@@ -0,0 +1,2136 @@
+ *   for advanced things like reallocating the memory of an entry (for the
+ *   purpose of defragmentation) and updating the pointer to the entry inside
+ *   the hash table.
+ */
+size_t hashtableScanDefrag(hashtable *ht, size_t cursor, hashtableScanFunction fn, void *privdata, void *(*defragfn)(void *), int flags) {
+    if (hashtableSize(ht) == 0) return 0;
+
+    /* Prevent entries from being moved around during the scan call, as a
+     * side-effect of the scan callback. */
+    hashtablePauseRehashing(ht);
+
+    /* Flags. */
+    int emit_ref = (flags & HASHTABLE_SCAN_EMIT_REF);
+
+    if (!hashtableIsRehashing(ht)) {
+        /* Emit entries at the cursor index. */
+        size_t mask = expToMask(ht->bucket_exp[0]);
+        bucket *b = &ht->tables[0][cursor & mask];
+        do {
+            if (b->presence != 0) {
+                int pos;
+                for (pos = 0; pos < ENTRIES_PER_BUCKET; pos++) {
+                    if (isPositionFilled(b, pos)) {
+                        void *emit = emit_ref ? &b->entries[pos] : b->entries[pos];
+                        fn(privdata, emit);
+                    }
+                }
+            }
+            bucket *next = bucketNext(b);
+            if (next != NULL && defragfn != NULL) {
+                next = bucketDefrag(b, next, defragfn);
+            }
+            b = next;
+        } while (b != NULL);
+
+        /* Advance cursor. */
+        cursor = nextCursor(cursor, mask);
+    } else {
+        int table_small, table_large;
+        if (ht->bucket_exp[0] <= ht->bucket_exp[1]) {
+            table_small = 0;
+            table_large = 1;
+        } else {
+            table_small = 1;
+            table_large = 0;
+        }
+
+        size_t mask_small = expToMask(ht->bucket_exp[table_small]);
+        size_t mask_large = expToMask(ht->bucket_exp[table_large]);
+
+        /* Emit entries in the smaller table, if this index hasn't already been
+         * rehashed. */
+        size_t idx = cursor & mask_small;
+        if (table_small == 1 || ht->rehash_idx == -1 || idx >= (size_t)ht->rehash_idx) {
+            size_t used_before = ht->used[table_small];
+            bucket *b = &ht->tables[table_small][idx];
+            do {
+                if (b->presence) {
+                    for (int pos = 0; pos < ENTRIES_PER_BUCKET; pos++) {
+                        if (isPositionFilled(b, pos)) {
+                            void *emit = emit_ref ? &b->entries[pos] : b->entries[pos];
+                            fn(privdata, emit);
+                        }
+                    }
+                }
+                bucket *next = bucketNext(b);
+                if (next != NULL && defragfn != NULL) {
+                    next = bucketDefrag(b, next, defragfn);
+                }
+                b = next;
+            } while (b != NULL);
+            /* If any entries were deleted, fill the holes. */
+            if (ht->used[table_small] < used_before) {
```

Comment:
I guess we could, but it didn't seem necessary, because rehashing is only paused in iterator, scan and two-phase-pop. It's not expected that any other elements are deleted during these procedures. I suggest we add it later if we find that it's necessary.

`hashtablePauseRehashing` can actually be made static, because it's only used internally.